File: apps/mcp-server/src/keyword/permission-forecast.ts
/**
 * Permission forecasting and approval bundling for parse_mode responses.
 *
 * Analyzes the current mode and task prompt to predict which permission
 * classes will be needed during execution, and groups related actions
 * into logical approval bundles so users understand upcoming interruptions.
 *
 * All functions in this module are **pure** — no I/O, no side effects.
 */

import type { ApprovalBundle, Mode, PermissionClass, PermissionForecast } from './keyword.types';

// ─── Prompt‑signal patterns ────────────────────────────────────────

const SHIP_PATTERNS = /\b(ship|deploy|push|pr\b|pull\s*request|merge|release)\b/i;

const TEST_PATTERNS = /\b(tests?|lint|typecheck|format|check|verify|coverage)\b/i;

const INSTALL_PATTERNS = /\b(install|add\s+package|add\s+dependency|yarn\s+add|npm\s+install)\b/i;

const DELETE_PATTERNS = /\b(delete|remove|drop|reset|clean|destroy)\b/i;

const REVIEW_PATTERNS = /\b(review|comment|approve|feedback|pr\b|pull\s*request)\b/i;

// ─── Pre-defined bundles ───────────────────────────────────────────

const SHIP_BUNDLE: ApprovalBundle = {
  name: 'Ship changes',
  actions: ['git add', 'git commit', 'git push', 'gh pr create'],
  permissionClass: 'external',
  reason: 'Committing, pushing, and creating a PR requires repo-write and GitHub API access',
};

const TEST_BUNDLE: ApprovalBundle = {
  name: 'Run checks',
  actions: ['yarn test', 'yarn lint', 'yarn typecheck'],
  permissionClass: 'read-only',
  reason: 'Running tests and linters reads the codebase but does not modify it',
};

const INSTALL_BUNDLE: ApprovalBundle = {
  name: 'Install dependencies',
  actions: ['yarn install', 'yarn add'],
  permissionClass: 'network',
  reason: 'Package installation fetches from external registries',
};

const REVIEW_BUNDLE: ApprovalBundle = {
  name: 'Review PR',
  actions: ['gh pr review', 'gh pr comment'],
  permissionClass: 'external',
  reason: 'Reviewing or commenting on a PR uses the GitHub API',
};

// ─── Base permission classes per mode ──────────────────────────────

const MODE_BASE_CLASSES: Record<Mode, PermissionClass[]> = {
  PLAN: ['read-only'],
  ACT: ['read-only', 'repo-write'],
  EVAL: ['read-only'],
  AUTO: ['read-only', 'repo-write', 'external'],
};

// ─── Public API ────────────────────────────────────────────────────

/**
 * Generate a permission forecast for the given mode and prompt.
 *
 * Pure function — safe to call from any context.
 */
export function generatePermissionForecast(mode: Mode, prompt: string): PermissionForecast {
  const classes = new Set<PermissionClass>(MODE_BASE_CLASSES[mode]);
  const bundles: ApprovalBundle[] = [];

  // Prompt-signal enrichment
  if (SHIP_PATTERNS.test(prompt)) {
    classes.add('repo-write');
    classes.add('external');
    bundles.push(SHIP_BUNDLE);
  }

  if (TEST_PATTERNS.test(prompt)) {
    bundles.push(TEST_BUNDLE);
  }

  if (INSTALL_PATTERNS.test(prompt)) {
    classes.add('network');
    bundles.push(INSTALL_BUNDLE);
  }

  if (DELETE_PATTERNS.test(prompt)) {
    classes.add('destructive');
  }

  if (REVIEW_PATTERNS.test(prompt) && mode === 'EVAL') {
    classes.add('external');
    bundles.push(REVIEW_BUNDLE);
  }

  // ACT mode always includes a "code changes" implicit bundle when no
  // explicit bundles matched
  if (mode === 'ACT' && bundles.length === 0) {
    bundles.push({
      name: 'Code changes',
      actions: ['file edit', 'git add', 'git commit'],
      permissionClass: 'repo-write',
      reason: 'Implementation involves editing files and creating commits',
    });
  }

  const permissionClasses = sortPermissionClasses([...classes]);

  return {
    permissionClasses,
    approvalBundles: bundles,
    permissionSummary: buildSummary(mode, permissionClasses, bundles),
  };
}

// ─── Helpers ───────────────────────────────────────────────────────

const CLASS_ORDER: PermissionClass[] = [
  'read-only',
  'repo-write',
  'network',
  'destructive',
  'external',
];

function sortPermissionClasses(classes: PermissionClass[]): PermissionClass[] {
  return [...classes].sort((a, b) => CLASS_ORDER.indexOf(a) - CLASS_ORDER.indexOf(b));
}

function buildSummary(mode: Mode, classes: PermissionClass[], bundles: ApprovalBundle[]): string {
  const highest = classes[classes.length - 1];
  const bundleNames = bundles.map(b => b.name).join(', ');

  if (bundles.length === 0) {
    return `${mode} mode requires ${highest} permissions`;
  }

  return `${mode} mode requires ${highest} permissions — expected actions: ${bundleNames}`;
}

File: packages/claude-code-plugin/hooks/lib/permission_forecast.py
"""Permission forecast formatting for CodingBuddy plugin (#1418).

Formats permission forecast data from parse_mode MCP responses and
generates standalone forecasts for self-contained mode.

All public functions are pure — no I/O, no side effects.
"""

from __future__ import annotations

import re
from typing import Dict, List, Optional, Sequence


# ─── Prompt-signal patterns (mirrors MCP permission-forecast.ts) ─────

_SHIP_RE = re.compile(
    r"\b(ship|deploy|push|pr\b|pull\s*request|merge|release)\b", re.I
)
_TEST_RE = re.compile(
    r"\b(tests?|lint|typecheck|format|check|verify|coverage)\b", re.I
)
_INSTALL_RE = re.compile(
    r"\b(install|add\s+package|add\s+dependency|yarn\s+add|npm\s+install)\b", re.I
)
_DELETE_RE = re.compile(
    r"\b(delete|remove|drop|reset|clean|destroy)\b", re.I
)
_REVIEW_RE = re.compile(
    r"\b(review|comment|approve|feedback|pr\b|pull\s*request)\b", re.I
)

# Pre-defined bundles (mirrors MCP permission-forecast.ts)
_SHIP_BUNDLE: Dict[str, str] = {"name": "Ship changes", "permissionClass": "external"}
_TEST_BUNDLE: Dict[str, str] = {"name": "Run checks", "permissionClass": "read-only"}
_INSTALL_BUNDLE: Dict[str, str] = {
    "name": "Install dependencies",
    "permissionClass": "network",
}
_DELETE_BUNDLE: Dict[str, str] = {
    "name": "Delete files",
    "permissionClass": "destructive",
}
_REVIEW_BUNDLE: Dict[str, str] = {"name": "Review PR", "permissionClass": "external"}

# Canonical sort order for permission classes
_CLASS_ORDER = ["read-only", "repo-write", "network", "destructive", "external"]


# ─── Permission class definitions ────────────────────────────────────

# Map of permission class to compact icon+label
PERMISSION_CLASS_LABELS: Dict[str, str] = {
    "read-only": "read-only",
    "repo-write": "repo-write",
    "network": "network",
    "destructive": "destructive",
    "external": "external",
}

# ─── Standalone forecasts per mode ────────────────────────────────────

# Default permission classes per mode (mirrors MCP server MODE_BASE_CLASSES)
MODE_BASE_CLASSES: Dict[str, List[str]] = {
    "PLAN": ["read-only"],
    "ACT": ["read-only", "repo-write"],
    "EVAL": ["read-only"],
    "AUTO": ["read-only", "repo-write", "external"],
}

# Default approval bundles per mode for standalone
MODE_DEFAULT_BUNDLES: Dict[str, List[Dict[str, str]]] = {
    "PLAN": [],
    "ACT": [
        {"name": "Code changes", "permissionClass": "repo-write"},
    ],
    "EVAL": [],
    "AUTO": [
        {"name": "Code changes", "permissionClass": "repo-write"},
        {"name": "Ship changes", "permissionClass": "external"},
    ],
}


# ─── Public API ───────────────────────────────────────────────────────


def format_permission_forecast(
    permission_classes: Sequence[str],
    approval_bundles: Optional[Sequence[Dict[str, str]]] = None,
) -> str:
    """Format permission forecast data as a compact status line.

    Args:
        permission_classes: List of permission class names
            (e.g. ["read-only", "repo-write"]).
        approval_bundles: Optional list of bundle dicts, each with
            at least ``name`` and ``permissionClass`` keys.

    Returns:
        Compact one-line string, e.g.
        ``Permissions: repo-write (Code changes) | external (Ship changes)``

        Returns empty string when there are no permission classes
        or only "read-only" with no bundles.
    """
    if not permission_classes:
        return ""

    # Filter out read-only when it is the only class and there are no bundles
    non_readonly = [c for c in permission_classes if c != "read-only"]
    if not non_readonly and not approval_bundles:
        return ""

    parts: list[str] = []

    if approval_bundles:
        # Group bundles by permission class for compact display
        for bundle in approval_bundles:
            name = bundle.get("name", "")
            pclass = bundle.get("permissionClass", "")
            label = PERMISSION_CLASS_LABELS.get(pclass, pclass)
            parts.append(f"{label} ({name})")
    else:
        # No bundles — just list the non-readonly classes
        for pclass in non_readonly:
            label = PERMISSION_CLASS_LABELS.get(pclass, pclass)
            parts.append(label)

    if not parts:
        return ""

    return "Permissions: " + " | ".join(parts)


def format_permission_forecast_from_mcp(
    forecast: Optional[Dict],
) -> str:
    """Extract and format permission forecast from a parse_mode MCP response.

    NOTE: Reserved for future MCP integration — not yet called by production code.

    Args:
        forecast: The ``permissionForecast`` dict from parse_mode, or None.

    Returns:
        Compact status line string, or empty string if no forecast data.
    """
    if not forecast:
        return ""

    classes = forecast.get("permissionClasses", [])
    bundles = forecast.get("approvalBundles", [])

    return format_permission_forecast(classes, bundles if bundles else None)


def generate_standalone_forecast(mode: str, prompt: Optional[str] = None) -> str:
    """Generate a permission forecast with optional prompt-aware analysis.

    When *prompt* is provided, analyzes it for task signals (install,
    test, ship, delete, review) and enriches the forecast beyond the
    static mode defaults — mirroring MCP ``permission-forecast.ts``.

    Args:
        mode: Mode name (PLAN, ACT, EVAL, AUTO).
        prompt: Optional raw user prompt for pattern analysis.

    Returns:
        Compact status line string, or empty string for read-only modes.
    """
    mode_upper = mode.upper()
    classes = set(MODE_BASE_CLASSES.get(mode_upper, []))
    bundles: list[Dict[str, str]] = []

    if prompt:
        if _SHIP_RE.search(prompt):
            classes.add("repo-write")
            classes.add("external")
            bundles.append(_SHIP_BUNDLE)
        if _TEST_RE.search(prompt):
            bundles.append(_TEST_BUNDLE)
        if _INSTALL_RE.search(prompt):
            classes.add("network")
            bundles.append(_INSTALL_BUNDLE)
        if _DELETE_RE.search(prompt):
            classes.add("destructive")
            bundles.append(_DELETE_BUNDLE)
        if _REVIEW_RE.search(prompt) and mode_upper == "EVAL":
            classes.add("external")
            bundles.append(_REVIEW_BUNDLE)

    # ACT mode implicit bundle when no explicit bundles matched
    if mode_upper == "ACT" and not bundles:
        bundles = list(MODE_DEFAULT_BUNDLES.get("ACT", []))

    # Sort classes canonically
    sorted_classes = sorted(classes, key=lambda c: _CLASS_ORDER.index(c) if c in _CLASS_ORDER else 99)

    return format_permission_forecast(sorted_classes, bundles if bundles else None)
